GVFS/GVFS.Common/NuGetUpgrade/NuGetUpgrader.cs

```diff
@@ -492,6 +522,91 @@ private bool TryRecursivelyDeleteInstallerDirectory(out string error)
             return true;
         }
 
+        private IList<IPackageSearchMetadata> QueryFeed(bool firstAttempt)
+        {
+            try
+            {
+                return this.nuGetFeed.QueryFeedAsync(this.nuGetUpgraderConfig.PackageFeedName).GetAwaiter().GetResult();
+            }
+            catch (Exception ex) when (firstAttempt &&
+                                       this.IsAuthRelatedException(ex))
+            {
+                // If we fail to query the feed due to an authorization error, then it is possible we have stale
+                // credentials, or credentials without the correct scope. Re-aquire fresh credentials and try again.
+                EventMetadata data = CreateEventMetadata(ex);
+                this.tracer.RelatedWarning(data, "Failed to query feed due to unauthorized error. Re-acquiring new credentials and trying again.");
+
+                if (!this.TryRefreshCredentials(out string error))
+                {
+                    // If we were unable to re-acquire credentials, throw a new exception indicating that we tried to handle this, but were unable to.
+                    throw new Exception($"Failed to query the feed for upgrade packages due to: {ex.Message}, and was not able to re-acquire new credentials due to: {error}", ex);
+                }
+
+                // Now that we have re-acquired credentials, try again - but with the retry flag set to false.
+                return this.QueryFeed(firstAttempt: false);
+            }
+            catch (Exception ex)
+            {
+                EventMetadata data = CreateEventMetadata(ex);
+                string message = $"Error encountered when querying NuGet feed. Is first attempt: {firstAttempt}.";
+                this.tracer.RelatedWarning(data, message);
+                throw new Exception($"Failed to query the NuGet package feed due to error: {ex.Message}", ex);
+            }
+        }
+
+        private bool IsAuthRelatedException(Exception ex)
+        {
+            // In observation, we have seen either an HttpRequestException directly, or
+            // a FatalProtocolException wrapping an HttpRequestException when we are not able
+            // to auth against the NuGet feed.
+            System.Net.Http.HttpRequestException httpRequestException = null;
+            if (ex is System.Net.Http.HttpRequestException)
+            {
+                httpRequestException = ex as System.Net.Http.HttpRequestException;
+            }
+            else if (ex is FatalProtocolException &&
+                ex.InnerException is System.Net.Http.HttpRequestException)
+            {
+                httpRequestException = ex.InnerException as System.Net.Http.HttpRequestException;

```

Comment:
Same comment about using a hard-cast here.
Alternatively you could also use the new pattern matching syntax to avoid a double type check:

```csharp
else if (ex is FatalProtocolException &&
         ex.InnerException is System.Net.Http.HttpRequestException innerHttpEx)
{
    httpRequestException = innerHttpEx;
}
```